[FSDP2] Added `shard_placement_fn` only for `_FlatShard`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136606



cc @XilunWu @H-Huang @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Differential Revision: [D63366290](https://our.internmc.facebook.com/intern/diff/D63366290)

Commit:
[FSDP2] Added `shard_placement_fn` only for `_FlatShard`

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added `shard_placement_fn` only for `_FlatShard`"




cc XilunWu H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63366290).

weifengpy commented on test/distributed/_composable/fsdp/test_fully_shard_init.py:
```diff
@@ -993,5 +993,67 @@ def test_hsdp_broadcast_across_replicas(self):
+        for param, ref_param in zip(model.parameters(), ref_model.parameters()):
+            full_param = param.full_tensor()
+            # NOTE: We must view back from 1D to ND.
+            full_param = full_param.view(param.shape)
```
for understanding purpose, DTensor.shape is original shape, but DTensor.full_tensor() is flattened shape?

awgu replied:
yep unfortunately it is like that right now
because `_FlatShard` is inheriting from `Shard`, the implementation to convert to replicated is the same, which will gather along the shard dim (which is 0), so it will still be 1D after the all-gather

I am not sure what the right thing to do here is exactly

awgu replied:
I think that the `DTensor.shape` should be the original unflattened shape since we need to hold that info somewhere. If we assume that as axiomatic, then we need to implement the ops for redistribution to be correct for `_FlatShard`, or we need to raise errors that certain redistribution is not supported.

We can fill out the `_FlatShard` redistribution logic over time. I can spend some time on it too later, but we should prioritize understanding which direction we want to go.

yifuwang replied:
Shall we do the reshape here? https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_redistribute.py#L156

weifengpy replied:
thanks for explaining. totally understand this is for directional exploration

weifengpy commented on torch/distributed/_composable/fsdp/_fsdp_param_group.py:
```diff
@@ -361,6 +365,10 @@ def post_backward(self, *unused: Any):
+                else:
+                    continue
+                if isinstance(fsdp_param.fsdp_placement, _FlatShard):
+                    unsharded_grads[-1] = unsharded_grads[-1].view(-1)
```
seems `.view(-1)` is the core change at fsdp side. curious should `.view(-1)` live in fsdp or dtensor? I can imagine a case where `view(-1) is done inside `DTensor.to_local()`

weifengpy replied:
the question is long-term and non-blocking. totally understnad we want to unblock in short-term

awgu replied:
> where view(-1) is done inside DTensor.to_local()

This unsharded grad is normally `torch.Tensor`. I think we have to do this `.view(-1)` so that the downstream post-backward logic works out of the box (e.g. computing what the padded unsharded size is for reduce-scatter). Otherwise, we may need more refactoring.

Commit:
Update on "[FSDP2] Added `shard_placement_fn` only for `_FlatShard`"




cc XilunWu H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D63366290](https://our.internmc.facebook.com/intern/diff/D63366290)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63366290).

Commit:
Update on "[FSDP2] Added `shard_placement_fn` only for `_FlatShard`"




cc XilunWu H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D63366290](https://our.internmc.facebook.com/intern/diff/D63366290)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63366290).

Commit:
Update on "[FSDP2] Added `shard_placement_fn` only for `_FlatShard`"




cc XilunWu H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D63366290](https://our.internmc.facebook.com/intern/diff/D63366290)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63366290).
